File: java/SegTree.java
class SegTree {
  Node root;

  SegTree(int[] values) {
    root = buildNode(values, 0, values.length - 1);
  }

  private Node buildNode(int[] values, int beginIndex, int endIndex) {
    if (beginIndex == endIndex) {
      return new Node(beginIndex, endIndex, values[beginIndex], null, null);
    }

    int middleIndex = (beginIndex + endIndex) / 2;
    Node left = buildNode(values, beginIndex, middleIndex);
    Node right = buildNode(values, middleIndex + 1, endIndex);

    return new Node(beginIndex, endIndex, Math.min(left.minValue, right.minValue), left, right);
  }

  void update(int index, int value) {
    update(index, value, root);
  }

  private void update(int index, int value, Node node) {
    if (node.beginIndex <= index && node.endIndex >= index) {
      if (node.beginIndex == node.endIndex) {
        node.minValue = value;
      } else {
        update(index, value, node.left);
        update(index, value, node.right);

        node.minValue = Math.min(node.left.minValue, node.right.minValue);
      }
    }
  }

  int query(int beginIndex, int endIndex) {
    return query(beginIndex, endIndex, root);
  }

  private int query(int beginIndex, int endIndex, Node node) {
    if (node.beginIndex > endIndex || node.endIndex < beginIndex) {
      return Integer.MAX_VALUE;
    }
    if (node.beginIndex >= beginIndex && node.endIndex <= endIndex) {
      return node.minValue;
    }

    return Math.min(
        query(beginIndex, endIndex, node.left), query(beginIndex, endIndex, node.right));
  }

  static class Node {
    int beginIndex;
    int endIndex;
    int minValue;
    Node left;
    Node right;

    Node(int beginIndex, int endIndex, int minValue, Node left, Node right) {
      this.beginIndex = beginIndex;
      this.endIndex = endIndex;
      this.minValue = minValue;
      this.left = left;
      this.right = right;
    }
  }
}

File: rust/src/seg_tree.rs
struct SegTree {
    root: Node,
}

#[allow(dead_code)]
impl SegTree {
    fn new(values: &[i32]) -> Self {
        Self {
            root: Self::build_node(values, 0, values.len() - 1),
        }
    }

    fn build_node(values: &[i32], begin_index: usize, end_index: usize) -> Node {
        let mut node = Node::new(begin_index, end_index);

        if begin_index == end_index {
            node.min_value = values[begin_index];
        } else {
            let middle_index = (begin_index + end_index) / 2;
            node.left = Some(Box::new(Self::build_node(
                values,
                begin_index,
                middle_index,
            )));
            node.right = Some(Box::new(Self::build_node(
                values,
                middle_index + 1,
                end_index,
            )));

            node.pull();
        }

        node
    }

    fn update(&mut self, index: usize, value: i32) {
        Self::update_node(index, value, &mut self.root);
    }

    fn update_node(index: usize, value: i32, node: &mut Node) {
        if node.begin_index <= index && node.end_index >= index {
            if node.begin_index == node.end_index {
                node.min_value = value;
            } else {
                Self::update_node(index, value, node.left.as_mut().unwrap());
                Self::update_node(index, value, node.right.as_mut().unwrap());

                node.pull();
            }
        }
    }

    fn query(&self, begin_index: usize, end_index: usize) -> i32 {
        Self::query_node(begin_index, end_index, &self.root)
    }

    fn query_node(begin_index: usize, end_index: usize, node: &Node) -> i32 {
        if node.begin_index > end_index || node.end_index < begin_index {
            return i32::MAX;
        }
        if node.begin_index >= begin_index && node.end_index <= end_index {
            return node.min_value;
        }

        Self::query_node(begin_index, end_index, node.left.as_ref().unwrap()).min(Self::query_node(
            begin_index,
            end_index,
            node.right.as_ref().unwrap(),
        ))
    }
}

struct Node {
    begin_index: usize,
    end_index: usize,
    min_value: i32,
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
}

impl Node {
    fn new(begin_index: usize, end_index: usize) -> Self {
        Self {
            begin_index,
            end_index,
            min_value: 0,
            left: None,
            right: None,
        }
    }

    fn pull(&mut self) {
        self.min_value = self
            .left
            .as_ref()
            .unwrap()
            .min_value
            .min(self.right.as_ref().unwrap().min_value);
    }
}
